### PR TITLE
Sketch for tinygo support

### PR DESCRIPTION
**Pull request by RubenRBS**
It currently does not work and if I remove the -gcflags flag, no asm is
produced (Instead of the current "unrecognized error")

If anyone is able to get this working, please feel free to help :)

For #3969

### CODE REVIEW

**Comment by partouf** [comment]
can you commit the compilers/tinygo class

**Comment by partouf** [comment]
did some investigating:
* building without target is super slow and I guess unusable
* using `-target arduino` requires `avr-gcc` to be in `PATH` (I hacked this into `getDefaultExecOptions() `and that worked `execOptions.env.PATH = execOptions.env.PATH + path.delimiter + '/opt/compiler-explorer/avr/gcc-11.1.0/bin';`)
* the number of `cgroup_pids_max` in the nsjail options needs to increase some more, I upped it from 64 to 128 and that worked at least
* gcflags not recognized, userOptions should just go directly into the args: 
```
    override optionsForFilter(filters, outputFilename, userOptions) {
        return ['build', '-o', outputFilename, ...userOptions];
    }
```
* there's no `-S`, and the workdir (if you use `-work`) doesn't contain asm either
* this also means that the `postProcess()` function is obsolete and should probably use the default implementation
* for the objdump to work we would need to use an appropriate objdump for the target, I tried this for `-target arduino` with `/opt/compiler-explrorer/avr/gcc-11.1.0/bin/avr-objdump` and that seems to work - but this obviously needs a bunch of manual labor from us to mix and match supported targets with objdumps that we have access to

image

File: lib/compilers/tinygo.ts [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+        super(compilerInfo, env);
+        this.goroot = this.compilerProps(`compiler.${this.compiler.id}.goroot`);
+        // this.goarch = this.compilerProps(`compiler.${this.compiler.id}.goarch`);
+        // this.goos = this.compilerProps(`compiler.${this.compiler.id}.goos`);
+    }
+
+    convertNewGoL(code) {
+        const re = /^\s+(0[Xx]?[\dA-Za-z]+)?\s?(\d+)\s*\(([^:]+):(\d+)\)\s*([A-Z]+)(.*)/;
+        const reUnknown = /^\s+(0[Xx]?[\dA-Za-z]+)?\s?(\d+)\s*\(<unknown line number>\)\s*([A-Z]+)(.*)/;
+        const reFunc = /TEXT\s+[".]*(\S+)\(SB\)/;
```
Comment (matsjla) [inline lib/compilers/tinygo.ts:56]: Can be moved into top level + maybe give them a more descriptive name

File: lib/compilers/tinygo.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+    }
+
+    constructor(compilerInfo, env) {
+        super(compilerInfo, env);
+        this.goroot = this.compilerProps(`compiler.${this.compiler.id}.goroot`);
+        // this.goarch = this.compilerProps(`compiler.${this.compiler.id}.goarch`);
+        // this.goos = this.compilerProps(`compiler.${this.compiler.id}.goos`);
+    }
+
+    convertNewGoL(code) {
```
Comment (matsjla) [inline lib/compilers/tinygo.ts:53]: Needs parameter and return type

File: lib/compilers/tinygo.ts (hunk 1 above)
Comment (RubenRBS) [reply lib/compilers/tinygo.ts:56]: Will do so also for golang.js when I move it to ts, good call

**Review by tony710521 — APPROVED** [review]

**Pull request closed without merging**